lib/Sema/BuilderTransform.cpp

```diff
@@ -1175,6 +1204,35 @@ class BuilderClosureRewriter
             capturedThen.first, {capturedThen.second.front()}));
     ifStmt->setThenStmt(newThen);
 
+    // Look for a #available condition. If there is one, we need to check
+    // that the resulting type of the "then" does refer to any types that
+    // are unavailable in the enclosing context.
+    //
+    // Note that this is for staging in support for
+    if (auto availabilityCond = findAvailabilityCondition(ifStmt->getCond())) {
+      SourceLoc loc = availabilityCond->getStartLoc();
+      Type thenBodyType = solution.simplifyType(
+          solution.getType(target.captured.second[0]));
+      thenBodyType.findIf([&](Type type) {
+        auto nominal = type->getAnyNominal();
+        if (!nominal)
+          return false;
+
+        if (auto reason = TypeChecker::checkDeclarationAvailability(
+                              nominal, loc, dc)) {
+          // Note that the problem is with the function builder, not the body.
+          // This is for staging only. We want to disable #available in
+          // function builders that don't support this operation.
```

Comment:
This doesn't seem reasonable in general; function builders which don't propagate types structurally are probably fine.  Can we not diagnose that an unavailable type has escaped the context?

Reply:
We can, and you're right: I'll zap the comment when address the other comments in a follow-up PR. Later in 5.3, we'll turn this warning into an error.